From 160eeb2263bd83e567da73f720db66c12df4a8db Mon Sep 17 00:00:00 2001
From: Tadeo Kondrak <me@tadeo.ca>
Date: Wed, 23 Sep 2020 20:33:04 +0100
Subject: [PATCH] Fix aerc breaking every time the package is rebuilt.

On NixOS, the SHAREDIR changes on every rebuild to the package, but aerc
fills it in as part of the default config and then installs that config
to the users home folder. Fix this by not substituting @SHAREDIR@ in the
default config until runtime.
---
 Makefile         |  2 +-
 config/config.go | 17 +++++++++++++++++
 2 files changed, 18 insertions(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 77f5e61..99961f0 100644
--- a/Makefile
+++ b/Makefile
@@ -23,7 +23,7 @@ aerc: $(GOSRC)
 		-o $@

 aerc.conf: config/aerc.conf.in
-	sed -e 's:@SHAREDIR@:$(SHAREDIR):g' > $@ < config/aerc.conf.in
+	cat config/aerc.conf.in > $@

 debug: $(GOSRC)
 	GOFLAGS="-tags=notmuch" \
diff --git a/config/config.go b/config/config.go
index 87d183a..6798059 100644
--- a/config/config.go
+++ b/config/config.go
@@ -470,6 +470,11 @@ func LoadConfigFromFile(root *string, sharedir string) (*AercConfig, error) {
 			return nil, err
 		}
 	}
+	if sec, err := file.GetSection("templates"); err == nil {
+		if key, err := sec.GetKey("template-dirs"); err == nil {
+			sec.NewKey("template-dirs", strings.ReplaceAll(key.String(), "@SHAREDIR@", sharedir))
+		}
+	}
 	file.NameMapper = mapName
 	config := &AercConfig{
 		Bindings: BindingConfig{
@@ -547,6 +552,18 @@ func LoadConfigFromFile(root *string, sharedir string) (*AercConfig, error) {
 		return nil, err
 	}

+	for i, filter := range config.Filters {
+		config.Filters[i].Command = strings.ReplaceAll(filter.Command, "@SHAREDIR@", sharedir)
+	}
+
+	for i, templateDir := range config.Templates.TemplateDirs {
+		config.Templates.TemplateDirs[i] = strings.ReplaceAll(templateDir, "@SHAREDIR@", sharedir)
+	}
+
+	for i, styleSetDir := range config.Ui.StyleSetDirs {
+		config.Ui.StyleSetDirs[i] = strings.ReplaceAll(styleSetDir, "@SHAREDIR@", sharedir)
+	}
+
 	if ui, err := file.GetSection("general"); err == nil {
 		if err := ui.MapTo(&config.General); err != nil {
 			return nil, err
--
2.29.2
